pkg/githubauth/app.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package githubauth
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/abcxyz/pkg/githubapp"
+)
+
+// AppScope is an interface which determines the repositories scope for which
+// the app should request tokens.
+type AppScope interface {
+	*allReposScope | *selectedReposScope
+}
+
+type allReposScope struct{}
+
+// ForAllRepos creates a scope that requests permissions on all repositories.
+func ForAllRepos() *allReposScope {
+	return &allReposScope{}
+}
+
+type selectedReposScope struct {
+	repos []string
+}
+
+// ForRepos creates a scope that requests permissions on the given repositories.
+func ForRepos(r ...string) *selectedReposScope {
+	return &selectedReposScope{
+		repos: r,
+	}
+}
+
+var _ TokenSource = (*AppTokenSource[*allReposScope])(nil)
+
+// AppTokenSource is a GitHubToken provider that authenticates as a GitHub App.
+type AppTokenSource[T AppScope] struct {
+	app   *githubapp.GitHubApp
+	scope T
+}
+
+// NewAppTokenSource returns a [AppTokenSource] which authenticates as a GitHub
+// App and returns a GitHub token.
+func NewAppTokenSource[T AppScope](appID, installationID, privateKeyPEM string, scope T) (*AppTokenSource[T], error) {
+	var merr error
+	if appID == "" {
+		merr = errors.Join(merr, fmt.Errorf("missing appID"))
+	}
+	if installationID == "" {
+		merr = errors.Join(merr, fmt.Errorf("missing installationID"))
+	}
+	if privateKeyPEM == "" {
+		merr = errors.Join(merr, fmt.Errorf("missing privateKeyPEM"))
+	}
+	if merr != nil {
+		return nil, fmt.Errorf("failed to create app token source: %w", merr)
+	}
+
+	privateKey, err := parseRSAPrivateKeyPEM(privateKeyPEM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create app token source: %w", err)
+	}
+
+	app := githubapp.New(githubapp.NewConfig(appID, installationID, privateKey))
+
+	return &AppTokenSource[T]{
+		app:   app,
+		scope: scope,
+	}, nil
+}
+
+func parseRSAPrivateKeyPEM(data string) (*rsa.PrivateKey, error) {
+	block, _ := pem.Decode([]byte(data))
+	if block == nil {
+		return nil, fmt.Errorf("failed to parse pem: no data")
```

Comment:
Maybe "failed to parse pem: no PEM block found". No data makes it sound like the data input was empty, which isn't necessarily the case.

Reply:
Done.